homeassistant/components/sunweg/__init__.py

```diff
@@ -0,0 +1,178 @@
+"""The Sun WEG inverter sensor integration."""
+import datetime
+import json
+import logging
+
+from sunweg.api import APIHelper
+from sunweg.plant import Plant
+
+from homeassistant import config_entries
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.util import Throttle
+
+from .const import CONF_PLANT_ID, DOMAIN, PLATFORMS
+
+SCAN_INTERVAL = datetime.timedelta(minutes=5)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: config_entries.ConfigEntry
+) -> bool:
+    """Load the saved entities."""
+    config = {**entry.data}
+    api = APIHelper(config[CONF_USERNAME], config[CONF_PASSWORD])
+    if not await hass.async_add_executor_job(api.authenticate):
+        _LOGGER.error("Username or Password may be incorrect!")
+        return False
+    if DOMAIN not in hass.data.keys():
+        hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][entry.entry_id] = SunWEGData(api, config[CONF_PLANT_ID])
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+
+class SunWEGData:
+    """The class for handling data retrieval."""
+
+    def __init__(
+        self,
+        api: APIHelper,
+        plant_id: int,
+    ) -> None:
+        """Initialize the probe."""
+
+        self.api = api
+        self.plant_id = plant_id
+        self.data: Plant = None
+        self.previous_values: dict = {}
```

Comment:
best practice is to type this dict.

Reply:
I can't. If I type it, I'll have type issues with a float(previous_value) below. The issue happens because of the datetime possible return.